### ISSUE TITLE
When you have a big database the m4m.mailchimp_sync_delta is null queries are very slow

### ISSUE THREAD

**Issue by maronk** [mailchimp/mc-magento#665]
select COUNT(*) from aha_162_production.magentomailchimp_ecommerce_sync_data;  returns close to 900K records for us.

In my understanding the is null query is slow in mysql for large tables - the following line generates a query that takes over 10 seconds to run every 5 minutes.

https://github.com/mailchimp/mc-magento/blob/f432a7d9516dce3d6b4431d452c81327e015c099/app/code/community/Ebizmarts/MailChimp/Model/Api/Customers.php#L288

By changing the above line from:
$collection->getSelect()->where("m4m.mailchimp_sync_delta IS null OR m4m.mailchimp_sync_modified = 1");

to:

$collection->getSelect()->where("(m4m.mailchimp_sync_delta < 0 AND m4m.mailchimp_sync_delta IS null) OR m4m.mailchimp_sync_modified = 1");

essentially we limit the potential results from the table since we remove all actual dates before checking for null speeding up the process a lot, this can probably be solved better in other ways and I am still testing this, but please consider whether there are ways to optimize this in a better way.

We have a lot of other queries in the mysql slow log from this extension with the same is null where statement.

Note that I am not an sql specialist, so feel free to criticise the approach and come up with a better one.

**Comment by Santiagoebizmarts** [comment]
Hello @maronk 

Thanks for your feedback on this. We will check this out and see how we can improve the queries searching for null in the mailchimp_sync_delta.

Regards.

**Comment by maronk** [comment]
Hello @Santiagoebizmarts 

My solution above is incorrect and generates inconsistent results - my apologies for this - after doing further research and understanding the aim of the code the problem is as follows:

The code currently generates this SQL

`
SELECT `e`.*, `at_prefix`.`value` AS `prefix`, `at_firstname`.`value` AS `firstname`, `at_middlename`.`value` AS `middlename`, `at_lastname`.`value` AS `lastname`, `at_suffix`.`value` AS `suffix`, 
CONCAT_WS(' ', IF(at_prefix.value IS NOT NULL AND at_prefix.value != '', 
LTRIM(RTRIM(at_prefix.value)), ''),
LTRIM(RTRIM(at_firstname.value)), 
IF(at_middlename.value IS NOT NULL AND at_middlename.value != '',
LTRIM(RTRIM(at_middlename.value)), ''), LTRIM(RTRIM(at_lastname.value)), 
IF(at_suffix.value IS NOT NULL AND at_suffix.value != '', 
LTRIM(RTRIM(at_suffix.value)), '')) AS `name`, `at_default_billing`.`value` AS `default_billing`, 
`at_postcode`.`value` AS `postcode`, `at_city`.`value` AS `city`, `at_region`.`value` AS `region`, 
`at_region_id`.`value` AS `region_id`, `at_country_id`.`value` AS `country_id`, `at_street`.`value` AS `street`,
`at_company`.`value` AS `company`, SUM(s.grand_total) AS total_spent, COUNT(s.entity_id) AS orders_count,
`m4m`.`related_id`, `m4m`.`type`, `m4m`.`mailchimp_store_id`, `m4m`.`mailchimp_sync_delta`, 
`m4m`.`mailchimp_sync_modified` FROM aha_162_production.magentocustomer_entity AS `e`
 LEFT JOIN aha_162_production.magentocustomer_entity_varchar AS `at_prefix` ON (`at_prefix`.`entity_id` = `e`.`entity_id`) AND (`at_prefix`.`attribute_id` = '4')
 LEFT JOIN aha_162_production.magentocustomer_entity_varchar AS `at_firstname` ON (`at_firstname`.`entity_id` = `e`.`entity_id`) AND (`at_firstname`.`attribute_id` = '5')
 LEFT JOIN aha_162_production.magentocustomer_entity_varchar AS `at_middlename` ON (`at_middlename`.`entity_id` = `e`.`entity_id`) AND (`at_middlename`.`attribute_id` = '6')
 LEFT JOIN aha_162_production.magentocustomer_entity_varchar AS `at_lastname` ON (`at_lastname`.`entity_id` = `e`.`entity_id`) AND (`at_lastname`.`attribute_id` = '7')
 LEFT JOIN aha_162_production.magentocustomer_entity_varchar AS `at_suffix` ON (`at_suffix`.`entity_id` = `e`.`entity_id`) AND (`at_suffix`.`attribute_id` = '8')
 LEFT JOIN aha_162_production.magentocustomer_entity_int AS `at_default_billing` ON (`at_default_billing`.`entity_id` = `e`.`entity_id`) AND (`at_default_billing`.`attribute_id` = '13')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_varchar AS `at_postcode` ON (`at_postcode`.`entity_id` = `at_default_billing`.`value`) AND (`at_postcode`.`attribute_id` = '29')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_varchar AS `at_city` ON (`at_city`.`entity_id` = `at_default_billing`.`value`) AND (`at_city`.`attribute_id` = '25')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_varchar AS `at_region` ON (`at_region`.`entity_id` = `at_default_billing`.`value`) AND (`at_region`.`attribute_id` = '27')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_int AS `at_region_id` ON (`at_region_id`.`entity_id` = `at_default_billing`.`value`) AND (`at_region_id`.`attribute_id` = '28')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_varchar AS `at_country_id` ON (`at_country_id`.`entity_id` = `at_default_billing`.`value`) AND (`at_country_id`.`attribute_id` = '26')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_text AS `at_street` ON (`at_street`.`entity_id` = `at_default_billing`.`value`) AND (`at_street`.`attribute_id` = '24')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_varchar AS `at_company` ON (`at_company`.`entity_id` = `at_default_billing`.`value`) AND (`at_company`.`attribute_id` = '23')
 LEFT JOIN aha_162_production.magentosales_flat_order AS `s` ON e.entity_id = s.customer_id
 LEFT JOIN aha_162_production.magentomailchimp_ecommerce_sync_data AS `m4m` ON m4m.related_id = e.entity_id and m4m.type = 'CUS' AND m4m.mailchimp_store_id = '17c9e74790a3c92ce75a6ae85f0945ca' and m4m.mailchimp_sync_delta is null
 WHERE (`e`.`entity_type_id` = '1') 
 AND (`e`.`store_id` = '1') AND ( m4m.mailchimp_sync_delta IS null OR m4m.mailchimp_sync_modified = 1) 
 GROUP BY `e`.`entity_id` LIMIT 100;
`

The last LEFT JOIN is joining the magentomailchimp_ecommerce_sync_data as m4m - then in the where statement we have an is null check which if I understand correctly is to return only those records that are missing from the magentomailchimp_ecommerce_sync_data table.

Since this join is done after all the others it seems like it is going through the full result set, joining all tables together, before excluding records at the end.

In an SQL editor, by moving the last SQL join first the time of the query is shortened from 8.6 seconds to 1.137 seconds.

I assume this way the sql engine first limits itself to the specific rows we want before joining all the other tables to the result set.

after the change the query looks as follows

`
SELECT `e`.*, `at_prefix`.`value` AS `prefix`, `at_firstname`.`value` AS `firstname`, `at_middlename`.`value` AS `middlename`, `at_lastname`.`value` AS `lastname`, `at_suffix`.`value` AS `suffix`, 
CONCAT_WS(' ', IF(at_prefix.value IS NOT NULL AND at_prefix.value != '', 
LTRIM(RTRIM(at_prefix.value)), ''),
LTRIM(RTRIM(at_firstname.value)), 
IF(at_middlename.value IS NOT NULL AND at_middlename.value != '',
LTRIM(RTRIM(at_middlename.value)), ''), LTRIM(RTRIM(at_lastname.value)), 
IF(at_suffix.value IS NOT NULL AND at_suffix.value != '', 
LTRIM(RTRIM(at_suffix.value)), '')) AS `name`, `at_default_billing`.`value` AS `default_billing`, 
`at_postcode`.`value` AS `postcode`, `at_city`.`value` AS `city`, `at_region`.`value` AS `region`, 
`at_region_id`.`value` AS `region_id`, `at_country_id`.`value` AS `country_id`, `at_street`.`value` AS `street`,
`at_company`.`value` AS `company`, SUM(s.grand_total) AS total_spent, COUNT(s.entity_id) AS orders_count,
`m4m`.`related_id`, `m4m`.`type`, `m4m`.`mailchimp_store_id`, `m4m`.`mailchimp_sync_delta`, 
`m4m`.`mailchimp_sync_modified` 
FROM aha_162_production.magentocustomer_entity AS `e`
 LEFT JOIN aha_162_production.magentomailchimp_ecommerce_sync_data AS `m4m` ON m4m.related_id = e.entity_id and m4m.type = 'CUS' AND m4m.mailchimp_store_id = '17c9e74790a3c92ce75a6ae85f0945ca' -- and m4m.mailchimp_sync_delta is null
 LEFT JOIN aha_162_production.magentocustomer_entity_varchar AS `at_prefix` ON (`at_prefix`.`entity_id` = `e`.`entity_id`) AND (`at_prefix`.`attribute_id` = '4')
 LEFT JOIN aha_162_production.magentocustomer_entity_varchar AS `at_firstname` ON (`at_firstname`.`entity_id` = `e`.`entity_id`) AND (`at_firstname`.`attribute_id` = '5')
 LEFT JOIN aha_162_production.magentocustomer_entity_varchar AS `at_middlename` ON (`at_middlename`.`entity_id` = `e`.`entity_id`) AND (`at_middlename`.`attribute_id` = '6')
 LEFT JOIN aha_162_production.magentocustomer_entity_varchar AS `at_lastname` ON (`at_lastname`.`entity_id` = `e`.`entity_id`) AND (`at_lastname`.`attribute_id` = '7')
 LEFT JOIN aha_162_production.magentocustomer_entity_varchar AS `at_suffix` ON (`at_suffix`.`entity_id` = `e`.`entity_id`) AND (`at_suffix`.`attribute_id` = '8')
 LEFT JOIN aha_162_production.magentocustomer_entity_int AS `at_default_billing` ON (`at_default_billing`.`entity_id` = `e`.`entity_id`) AND (`at_default_billing`.`attribute_id` = '13')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_varchar AS `at_postcode` ON (`at_postcode`.`entity_id` = `at_default_billing`.`value`) AND (`at_postcode`.`attribute_id` = '29')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_varchar AS `at_city` ON (`at_city`.`entity_id` = `at_default_billing`.`value`) AND (`at_city`.`attribute_id` = '25')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_varchar AS `at_region` ON (`at_region`.`entity_id` = `at_default_billing`.`value`) AND (`at_region`.`attribute_id` = '27')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_int AS `at_region_id` ON (`at_region_id`.`entity_id` = `at_default_billing`.`value`) AND (`at_region_id`.`attribute_id` = '28')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_varchar AS `at_country_id` ON (`at_country_id`.`entity_id` = `at_default_billing`.`value`) AND (`at_country_id`.`attribute_id` = '26')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_text AS `at_street` ON (`at_street`.`entity_id` = `at_default_billing`.`value`) AND (`at_street`.`attribute_id` = '24')
 LEFT JOIN aha_162_production.magentocustomer_address_entity_varchar AS `at_company` ON (`at_company`.`entity_id` = `at_default_billing`.`value`) AND (`at_company`.`attribute_id` = '23')
 LEFT JOIN aha_162_production.magentosales_flat_order AS `s` ON e.entity_id = s.customer_id
 WHERE (`e`.`entity_type_id` = '1') 
 AND (`e`.`store_id` = '1') AND ( m4m.mailchimp_sync_delta IS null OR m4m.mailchimp_sync_modified = 1) 
 GROUP BY `e`.`entity_id` LIMIT 100;
`

Is there some way we can force this join to be the first - this will improve the performance significantly - this would apply both with products and customers.

**Comment by seansan** [comment]
is sync delta every null? also it doesnt default to null when empty

![image](https://user-images.githubusercontent.com/652395/37464658-4e460d84-2859-11e8-93cb-82baa0dc4815.png)

**Comment by maronk** [comment]
@seansan 
No - it is never null.  The left join for null records is to find out which records do not exist in the m4m table - when you join like this that field will become null in the result set, even if null is not allowed.

This means, that if you join this one last, all the other joins will happen for the whole resultset, if you put the join first, only the rows which do not exist in the m4m table will need to be joined with the attributes.  Still a slow query but about 8x faster than the current.

**Comment by centerax** [comment]
the magic is to create the query using magento code not raw sql

**Comment by maronk** [comment]
@Santiagoebizmarts @centerax   I'm sorry if I'm not of much help when it comes to the magento code, but when it comes to the code in magento that is not my strong side - however the sql suggested above consistently fixes the issue.

Our database server consistently goes to 100% cpu when they bulksync operations start, which is causing crashes, especially when the load on the frontend is high when the queries start running.  To solve this I have changed our config so that the bulksync operations are only running on slower times of the day.  From app/code/community/Ebizmarts/MailChimp/etc/config.xml

Hope this can help others until the underlying issue is solved.

`
     <mailchimp_bulksync_ecommerce_data>
                <schedule>
                    <cron_expr>*/5 0,1,2,3,5,22,23 * * *</cron_expr>
                </schedule>
                <run>
                    <model>mailchimp/cron::syncEcommerceBatchData</model>
                </run>
            </mailchimp_bulksync_ecommerce_data>
            <mailchimp_bulksync_subscriber_data>
                <schedule>
                    <cron_expr>*/5 0,1,2,3,5,22,23 * * *</cron_expr>
                </schedule>
                <run>
                    <model>mailchimp/cron::syncSubscriberBatchData</model>
                </run>
            </mailchimp_bulksync_subscriber_data>
`

**Comment by Santiagoebizmarts** [comment]
@maronk 

Thanks again, we believe performance improvements to be very importante.
We will do our best to address this as soon as possible.

Regards.

**Comment by bizlift** [comment]
Any ETA on this?  The sync cron jobs frequently crash even though config is set to limit of 50.  Even after bumping up PHP max memory & execution time.

**Comment by sprankhub** [comment]
Bump! The cronjob `mailchimp_bulksync_ecommerce_data` crashes for us frequently as well.

**Comment by TroyPatteson** [comment]
The performance of this query has been an issue for the magento stores we have as well. We resolved this issue by refactoring the query into two. We are running a patched version of mc-magento v1.1.7 so I can't provide a direct pull request but I thought it might be useful to describe our solution. Apologies in advance for the length of this post.

If you look at the query above there are really two parts to it. One part is identifying the next 100 customers that need to be synchronised. The second part is getting the customer data required to synchronise.

The first part is only dependent on two tables: <i>customer_entity</i> and <i>mailchimp_ecommerce_sync_data</i>. A customer record needs to be synchronised if it hasn't been before (left join on <i>mailchimp_ecommerce_sync_data</i> returns NULL) or if it has been modified since it was last synchronised (mailchimp_sync_modified = 1).

The second part joins the primary billing address EAV attributes (if any) and then joins all orders for the given customer to be able to get the total spent. Since each customer can have more than one order, the selected rows must be grouped on customer <i>entity_id</i> and the order totals summed to get the total spent. It is this grouping that causes the problem. Effectively every customer must be retrieved and joined with all of the primary billing address EAVs and orders <b>before</b> the first 100 rows can be selected.

The solution is to split the query to first get the 100 customer entity IDs to synchronise and then perform the second part of the query on <b>only</b> those 100 rows to get the customer data.

Based on the v1.1.7 code base the changes required are to add a new method to get the customers to sync and modify method <i>makeCustomersNotSentCollection()</i> to use those customers as follows:

```
    /**
     * Get an array of customer entity IDs of the next batch of customers
     * to sync.
     * @return int[] Customer IDs to sync
     * @throws Mage_Core_Exception
     */
    protected function getCustomersToSync()
    {
        $collection = $this->getCustomerResourceCollection();
        $collection->getSelect()->reset(Zend_Db_Select::COLUMNS)->columns($collection->getRowIdFieldName());
        $collection->addFieldToFilter('website_id', array('eq' => $this->getWebsiteIdForStoreId($this->getBatchMagentoStoreId())));
        $this->joinMailchimpSyncData($collection);
        $collection->getSelect()->limit($this->getBatchLimitFromConfig());

        $customerIds = array();
        foreach ($collection as $customer) {
            array_push($customerIds, $customer->getId());
        }
        return $customerIds;
    }

    /**
     * @param int[] $customerIdsToSync Customer IDs to synchronise.
     * @return Mage_Customer_Model_Resource_Customer_Collection
     */
    public function makeCustomersNotSentCollection($customerIdsToSync)
    {
        /**
         * @var Mage_Customer_Model_Resource_Customer_Collection $collection
         */
        $collection = $this->getCustomerResourceCollection();
        $collection->addFieldToFilter('entity_id', array('in' => $customerIdsToSync));
        $collection->addNameToSelect();
        $this->joinDefaultBillingAddress($collection);
        $this->joinSalesData($collection);
        $collection->getSelect()->group("e.entity_id");
        return $collection;
    }
````

Note that we only return customer entity IDs limited by the maximum batch size in <i>getCustomersToSync()</i>. There is no point returning other columns that we don't need. As a result we also modify method <i>joinMailchimpSyncData()</i> to not return any columns:

```
    /**
     * @param Mage_Customer_Model_Resource_Customer_Collection $collection
     */
    protected function joinMailchimpSyncData($collection)
    {
        $joinCondition      = "m4m.related_id = e.entity_id and m4m.type = '%s' AND m4m.mailchimp_store_id = '%s'";
        $mailchimpTableName = $this->getSyncdataTableName();

        $collection->getSelect()->joinLeft(
            array("m4m" => $mailchimpTableName),
            sprintf($joinCondition, Ebizmarts_MailChimp_Model_Config::IS_CUSTOMER, $this->mailchimpStoreId),
            array()
        );

        $collection->getSelect()->where("m4m.mailchimp_sync_delta IS null OR m4m.mailchimp_sync_modified = 1");
    }
```

Finally, we modify <i>createBatchJson()</i> to call the new method as follows:

```    public function createBatchJson($mailchimpStoreId, $magentoStoreId)
    {
        $this->mailchimpStoreId = $mailchimpStoreId;
        $this->magentoStoreId   = $magentoStoreId;

        $collection = array();
        $customerIds = $this->getCustomersToSync();
        if (count($customerIds) > 0) {
            $collection = $this->makeCustomersNotSentCollection($customerIds);
        }

        $customerArray = array();
        
        $this->makeBatchId();

        ....
    }
```

On our largest website we have around 1 million customers and 2.85 million rows in the <i>mailchimp_ecommerce_sync_data</i> table.

Before the changes the query would take around 100 seconds to run. After the changes the two queries together would take around 10 seconds to run.

It shouldn't be too hard to adapt these changes to new code base. There may also be scope to adopt a similar approach for other queries.

**Comment by Santiagoebizmarts** [comment]
Hello @TroyPatteson 

Thank you very much for this input. It does make a lot of sense what you mention.
We've seen some problems with the '**in**' filter when the arrays included are too large. We will run a couple tests and see if we will apply this exactly or do a couple of modifications to avoid possible errors.

This is a great contribution, we really appreciate it!

Kind Regards.
